### PR TITLE
sae: Implement custom tx `AsOp`

### PR DESCRIPTION
**Pull request by StephenButtolph**
## Why this should be merged

This PR factors out `Tx.AsOp` from #5303.

## How this works

`Tx.AsOp` includes multiple pieces of critical data for SAE:
- [X] `op.Gas` - which correlates to `atomic.UnsignedTx.GasUsed` in coreth.
- [X] `op.GasFeeCap` - which correlates to `atomic.EffectiveGasPrice` in coreth.
- [X] `op.Burn` and `op.Mint` - which correlate to the non-multicoin modifications in `atomic.UnsignedTx.EVMStateTransfer` in coreth.

## How this was tested

- [X] Extended existing mainnet tx tests to include ops
- [X] Extended tx tests to include additional mainnet tx
- [X] Extended tx tests to include synthetic txs for edge cases
- [X] Added `TestAsOp_Errors` to test error conditions
- [X] Fuzzed `AsOp` against coreth

## Need to be documented in RELEASES.md?

No

### CODE REVIEW

File: vms/saevm/cchain/tx/tx.go [hunk 1]
```diff
@@ -64,6 +85,98 @@ func (t *Tx) Bytes() ([]byte, error) {
 	return c.Marshal(codecVersion, t)
 }
 
+// AsOp converts the transaction into a [hook.Op] that can be processed by SAE.
+//
+// The operation only includes state changes that impact Ethereum-native state.
+// It does not include non-AVAX balance changes or shared memory modifications.
+func (t *Tx) AsOp(avaxAssetID ids.ID) (hook.Op, error) {
```
Comment (StephenButtolph) [inline vms/saevm/cchain/tx/tx.go:92]: This is the only exported function added in this PR.

File: vms/saevm/cchain/tx/export.go [hunk 2]
```diff
@@ -4,10 +4,15 @@
...
+
 	"github.com/ava-labs/libevm/common"
 
 	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/math"
 	"github.com/ava-labs/avalanchego/vms/components/avax"
+	"github.com/ava-labs/avalanchego/vms/saevm/hook"
 )
 
 // Export is the unsigned component of a transaction that transfers assets from
```
Comment (alarso16) [inline vms/saevm/cchain/tx/export.go:23]: You don't actually assert that the types implement the interface. Any reason not to add it?
```suggestion
var _ Unsigned = (*Export)(nil)

// Export is the unsigned component of a transaction that transfers assets from
```

And the same for import

File: vms/saevm/cchain/tx/export.go [hunk 3]
```diff
@@ -36,3 +38,59 @@ type Input struct {
...
 }
+
+func (e *Export) burned(assetID ids.ID) (uint64, error) {
+	var (
+		burned uint64
+		err    error
+	)
+	for _, in := range e.Ins {
+		if in.AssetID == assetID {
+			burned, err = math.Add(burned, in.Amount)
```
Comment (alarso16) [inline vms/saevm/cchain/tx/export.go:60]: A somewhat pedantic question, as we've already discussed about how the supply is guaranteed to prevent overflows like this, but there is a chance that we can return a valid `uint64` despite an overflow here, because the intermediate value overflowed before the subtractions below. Does this matter?

File: vms/saevm/cchain/tx/tx.go [hunk 4]
```diff
@@ -62,6 +85,98 @@ func (t *Tx) Bytes() ([]byte, error) {
...
+
+const _x2cRate = 1_000_000_000
+
+// x2cRate is the conversion rate between the smallest denomination on the
+// X-Chain, 1 nAVAX, and the smallest denomination on the C-Chain 1 aAVAX.
+var x2cRate = uint256.NewInt(_x2cRate)
+
+// scaleAVAX converts an amount denominated in nAVAX into the C-Chain's aAVAX
+// denomination.
+func scaleAVAX(nAVAX uint64) uint256.Int {
```
Comment (alarso16) [inline vms/saevm/cchain/tx/tx.go:162]: IMO this name should be clear that this is a one way conversion from X-chain to C-chain. I don't have a better name for this, but reading the code doesn't not make it clear the denominations we end up with

File: vms/saevm/cchain/tx/tx.go [hunk 5]
```diff
@@ -7,8 +7,20 @@
 package tx
 
 import (
+	"fmt"
+
+	"github.com/ava-labs/libevm/common"
+	"github.com/holiman/uint256"
+
+	// Imported for [atomic.TxBytesGas] comment resolution.
+	_ "github.com/ava-labs/avalanchego/graft/coreth/plugin/evm/atomic"
```
Comment (alarso16) [inline vms/saevm/cchain/tx/tx.go:16]: I don't think you should import this just for the comment resolution. You could write out the path

File: vms/saevm/cchain/tx/tx_test.go [hunk 6]
```diff
@@ -22,19 +23,23 @@ import (
...
 	"github.com/ava-labs/avalanchego/vms/secp256k1fx"
+
+	safemath "github.com/ava-labs/avalanchego/utils/math"
 )
 
 // Tests is defined at the package level to allow sharing between fuzz tests and
 // unit tests.
 var (
-	Tests = [...]struct {
+	AVAXAssetID = ids.FromStringOrPanic("FvwEAhmxKfeiG8SnEvq42hc6whRyY3EFYAvebMqDNDGCgxN5Z")
```
Comment (alarso16) [inline vms/saevm/cchain/tx/tx_test.go:36]: Is this really not already defined as a constant somewhere?

File: vms/saevm/cchain/tx/tx_test.go [hunk 7]
```diff
@@ -133,8 +138,14 @@ var (
 					]
 				}]
 			}`,
-			ID:    ids.FromStringOrPanic("h34BPNmYApCbW8buVWAtzu1KtjTFmyMhiRQQnAqPqwCqQsB7f"),
 			Bytes: common.FromHex("0x000000000000000000010427d4b22a2a78bcddd456742caf91b56badbff985ee19aef14573e7343fd652ed5f38341e436e5d46e2bb00b45d62ae97d1b050c64bc634ae10626739e35c4b00000001c52b712aa7dce27a650bf509f799673e245edd4fa9e4e1700eb6105202fe579a0000000121e67317cbc4be2aeb00677ad6462778a8f52274b9d605df2591b23027a87dff000000050000000002faf080000000010000000000000001b8b5a87d1c05676f1f966da49151fa54dbe68c330000000002faf08021e67317cbc4be2aeb00677ad6462778a8f52274b9d605df2591b23027a87dff0000000100000009000000013e6614876ee01d3b8b27480c00bdcb0ae84ee3e8346d2d5f08320f7dd3e76c4540be021fe85e91817654c9310b54e8f2e88d81db52b8693842b90f3dbd23bd5c01"),
+			Op: hook.Op{
+				ID:  ids.FromStringOrPanic("h34BPNmYApCbW8buVWAtzu1KtjTFmyMhiRQQnAqPqwCqQsB7f"),
+				Gas: 11230,
+				Mint: map[common.Address]uint256.Int{
+					common.HexToAddress("0xb8b5a87d1c05676f1f966da49151fa54dbe68c33"): *uint256.NewInt(50_000_000 * _x2cRate),
```
Comment (alarso16) [inline vms/saevm/cchain/tx/tx_test.go]:
```suggestion
					common.HexToAddress("0xb8b5a87d1c05676f1f966da49151fa54dbe68c33"): scaleAVAX(50_000_000),
```
You defined the function for this, right?

File: vms/saevm/cchain/tx/tx_test.go [hunk 8]
```diff
@@ -232,8 +243,421 @@ var (
...
+					},
+				},
+			},
+		},
+		{
+			Name: "import_multi_input", // Included in https://subnets.avax.network/c-chain/block/132481
+			Old: &atomic.Tx{
+				UnsignedAtomicTx: &atomic.UnsignedImportTx{
+					NetworkID:    1,
+					BlockchainID: ids.FromStringOrPanic("2q9e4r6Mu3U68nU1fYjgbR6JvwrRx36CohpAX5UQxse55x1Q5"),
```
Comment (alarso16) [inline vms/saevm/cchain/tx/tx_test.go:265]: Any reason you didn't abstract the chain IDs into a global? I guess it shouldn't matter much, but it might be more readable

File: vms/saevm/cchain/tx/tx_test.go [hunk 9]
```diff
@@ -22,19 +23,23 @@ import (
 	"github.com/ava-labs/avalanchego/vms/components/avax"
 	"github.com/ava-labs/avalanchego/vms/components/verify"
 	"github.com/ava-labs/avalanchego/vms/saevm/cmputils"
+	"github.com/ava-labs/avalanchego/vms/saevm/hook"
 	"github.com/ava-labs/avalanchego/vms/secp256k1fx"
+
+	safemath "github.com/ava-labs/avalanchego/utils/math"
```
Comment (alarso16) [inline vms/saevm/cchain/tx/tx_test.go:30]: You don't have to alias this - i have seen this aliased everywhere else though, so maybe it's "conventional" - but at least match `tx.go`

File: vms/saevm/cchain/tx/tx_test.go [hunk 10]
```diff
@@ -513,3 +937,77 @@ func TestJSONMarshal(t *testing.T) {
...
+	for _, test := range Tests {
+		t.Run(test.Name, func(t *testing.T) {
+			got, err := test.New.AsOp(AVAXAssetID)
+			require.NoErrorf(t, err, "%T.AsOp(AVAXAssetID)", test.New)
+			assert.Equalf(t, test.Op, got, "%T.AsOp(AVAXAssetID)", test.New)
+		})
+	}
+}
+
+func TestAsOp_Errors(t *testing.T) {
```
Comment (alarso16) [inline vms/saevm/cchain/tx/tx_test.go:951]: There's quite a few other overflow conditions that aren't tested

File: vms/saevm/cchain/tx/tx_test.go (hunk 8 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/tx_test.go:265]: I only pulled `AVAXAssetID` out because I needed it to pass into functions that I was testing. I didn't really think about whether or not other fields should be moved out just for the reuse...

I'm not really sure where that would end (for example, should we have a variable for any of these fields that are duplicated?).

I've done that in the past for when we were constructing txs like this... But I feel now that the tradeoff of knowing the tx structure is better than reducing the duplication... I could probably get convinced of reducing the duplication.

For example, `*avax.TransferableInput`, `*avax.TransferableOutput`, and `*secp256k1fx.Credential` are always exactly duplicated between `old` and `new`... We could define them as a single var and then share them between the txs.

File: vms/saevm/cchain/tx/tx_test.go (hunk 6 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/tx_test.go:36]: It doesn't seem to be. (At least not in a place that we can import it).

It exists in a couple other tests

File: vms/saevm/cchain/tx/export.go (hunk 3 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/export.go:60]: This is worth a comment, 🦅👁️.

File: vms/saevm/cchain/tx/export.go (hunk 2 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/export.go:23]: I'm fine with adding it

File: vms/saevm/cchain/tx/export.go (hunk 3 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/export.go:60]: I decided not to actually change the function, since it makes fuzzing quite a bit easier. Although we could change it in the future if we wanted to, the code can't be hit in practice.

File: vms/saevm/cchain/tx/tx.go (hunk 5 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/tx.go:16]: Hm, why not? Is the concern about `init` functions in coreth? I don't think there is anything too concerning in the init functions.. I suppose maybe the concern is around the ext data hashes?

File: vms/saevm/cchain/tx/tx.go (hunk 4 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/tx.go:162]: I suppose I was relying on the types to make it clear that the AVAX was being scaled upwards rather than downwards. `xAVAXToC` feels like a horrible name... `toAAVAX` also seems gross... `bigAVAX` could work, although I'm not sure that really addresses your concern

Comment (alarso16) [reply vms/saevm/cchain/tx/tx.go:162]: Not a big enough deal to block merge, and those options are worse... I guess we can keep it

File: vms/saevm/cchain/tx/tx.go (hunk 5 above)
Comment (alarso16) [reply vms/saevm/cchain/tx/tx.go:16]: meh whatever, i just thought it was unnecessary. but we can punt this until we eventually deprecate coreth completely

File: vms/saevm/cchain/tx/tx_test.go (hunk 8 above)
Comment (alarso16) [reply vms/saevm/cchain/tx/tx_test.go:265]: yeah a slippery slope for sure. writer discretion then, or if it gets worse later, we can always add it

**Review by alarso16 — APPROVED** [review]

File: vms/saevm/cchain/tx/tx.go [hunk 11]
```diff
@@ -62,6 +85,98 @@ func (t *Tx) Bytes() ([]byte, error) {
...
+
+const _x2cRate = 1_000_000_000
+
+// x2cRate is the conversion rate between the smallest denomination on the
+// X-Chain, 1 nAVAX, and the smallest denomination on the C-Chain 1 aAVAX.
+var x2cRate = uint256.NewInt(_x2cRate)
+
+// scaleAVAX converts an amount denominated in nAVAX into the C-Chain's aAVAX
+// denomination.
+func scaleAVAX(nAVAX uint64) uint256.Int {
```
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/tx.go:162]: I should have known this but I don't think I did -- why historically, did C-Chain use a different denomination?

File: vms/saevm/cchain/tx/tx.go (hunk 5 above)
Comment (JonathanOppenheimer) [reply vms/saevm/cchain/tx/tx.go:16]: I think we've been fine adding imports for comment resolution previously.

File: vms/saevm/cchain/tx/compatibility_test.go [hunk 12]
```diff
@@ -51,3 +59,96 @@ func FuzzJSONCompatibility(f *testing.F) {
 		assert.JSONEq(t, string(want), string(got))
 	})
 }
+
+func FuzzAsOpCompatibility(f *testing.F) {
+	fuzz(f, func(t *testing.T, newTx *Tx) {
+		got, err := newTx.AsOp(AVAXAssetID)
```
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/compatibility_test.go:65]: how were the the asset ID's chosen. there was no fun easter egg :(

File: vms/saevm/cchain/tx/tx.go [hunk 13]
```diff
@@ -62,6 +85,98 @@ func (t *Tx) Bytes() ([]byte, error) {
 	return c.Marshal(codecVersion, t)
 }
 
+// AsOp converts the transaction into a [hook.Op] that can be processed by SAE.
+//
+// The operation only includes state changes that impact Ethereum-native state.
+// It does not include non-AVAX balance changes or shared memory modifications.
+func (t *Tx) AsOp(avaxAssetID ids.ID) (hook.Op, error) {
```
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/tx.go:92]: Is `Tx` intended to grow beyond the atomic Import/Export txs -- this seems to me to be basically atomic txs type with a generic name.

File: vms/saevm/cchain/tx/tx.go (hunk 1 above)
Comment (JonathanOppenheimer) [reply vms/saevm/cchain/tx/tx.go:92]: josh is somewhere thinking about narrow api surfaces

File: vms/saevm/cchain/tx/compatibility_test.go [hunk 14]
```diff
@@ -51,3 +59,96 @@ func FuzzJSONCompatibility(f *testing.F) {
...
+		if err != nil {
+			t.Skip("invalid tx")
+		}
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/compatibility_test.go:68]: The code clearly compiles.

File: vms/saevm/cchain/tx/tx.go (hunk 13 above)
Comment (JonathanOppenheimer) [reply vms/saevm/cchain/tx/tx.go:92]: (I haven't looked closely at the big PR)

File: vms/saevm/cchain/tx/tx.go [hunk 15]
```diff
@@ -62,6 +85,98 @@ func (t *Tx) Bytes() ([]byte, error) {
...
+// gasPrice takes in the cost, in nAVAX, and the gas and returns the price per
+// gas in aAVAX/gas.
+//
+// The result is rounded down to the nearest aAVAX/gas.
+func gasPrice(cost uint64, gas gas.Gas) uint256.Int {
+	var u uint256.Int
+	u.SetUint64(uint64(gas))
+
+	p := scaleAVAX(cost)
+	p.Div(&p, &u)
```
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/tx.go:178]: comment this is guaranteed to be non-zero?

File: vms/saevm/cchain/tx/export.go [hunk 16]
```diff
@@ -36,3 +43,64 @@ type Input struct {
...
 	Nonce   uint64         `serialize:"true" json:"nonce"`
 }
+
+// Like [atomic.UnsignedExportTx.Burned], burned will error if the sum of the
+// inputs exceeds MaxUint64, even if the total amount burned could be
+// represented as a uint64.
+//
+// Because the total supply of AVAX fits in a uint64, this doesn't matter in
+// practice and allows for easier fuzzing.
+func (e *Export) burned(assetID ids.ID) (uint64, error) {
```
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/export.go:53]: This is practically identical to `Import` version -- and vice versa. Would it be worth to abstract this or nah?

Comment (JonathanOppenheimer) [reply vms/saevm/cchain/tx/export.go:53]: Tbh I think this is one of those cases where even though it's nearly identical the answer is no.

File: vms/saevm/cchain/tx/tx.go [hunk 17]
```diff
@@ -62,6 +85,98 @@ func (t *Tx) Bytes() ([]byte, error) {
...
+		return 0, err
+	}
+	dynamicGas, err := math.Add(bytesGas, sigsGas)
+	if err != nil {
+		return 0, err
+	}
+	return math.Add(intrinsicGas, dynamicGas)
+}
+
+const _x2cRate = 1_000_000_000
```
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/tx.go:154]: I feel like consts like this should all be declared in a single place somewhere, single source of truth style. 

We already have `avaxConversionRateInt` in the wallet for example.

Comment (JonathanOppenheimer) [reply vms/saevm/cchain/tx/tx.go:154]: in the vms package. might be out of scope for this PR, but seemed like an opportunity for improvement.

File: vms/saevm/cchain/tx/tx.go [hunk 18]
```diff
@@ -62,6 +85,98 @@ func (t *Tx) Bytes() ([]byte, error) {
...
+		Mint:      op.mint,
+	}, nil
+}
+
+const (
+	// intrinsicGas is an initial static amount of gas that every [Tx] must pay.
+	intrinsicGas = ap5.AtomicTxIntrinsicGas
+	// gasPerByte is an additional amount of gas that is charged per-byte of an
+	// [Unsigned] transaction.
+	gasPerByte = 1 // [atomic.TxBytesGas]
```
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/tx.go:122]: won't the [atomic] package get deleted? this comment will break. also why do we hardcode it here, but keep it as something nice in the atomic package.

File: vms/saevm/cchain/tx/export.go [hunk 19]
```diff
@@ -4,12 +4,22 @@
 package tx
 
 import (
+	"errors"
+	"fmt"
+
 	"github.com/ava-labs/libevm/common"
 
+	// Imported for [atomic.UnsignedExportTx.Burned] comment resolution.
```
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/export.go:12]: Is there a linter for these doc comments. 

1. I mess these up and Austin suggested I start a server locally.
2. When this package gets deleted this it will have to be fixed.

File: vms/saevm/cchain/tx/tx_test.go [hunk 20]
```diff
@@ -6,11 +6,13 @@ package tx
 import (
```
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/tx_test.go:6]: need to review and understand still

**Review by JonathanOppenheimer — COMMENTED** [review]
need to go through tests more

File: vms/saevm/cchain/tx/tx.go (hunk 11 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/tx.go:162]: When deciding the denomination of AVAX on the X-chain and the P-chain, we decided to make it divisible into nanos (10^-9) since that was the most it could be broken up while still having the total supply fit into a uint64.

On the C-chain, we are wed to how ethereum did it, which is having Eth be fractional to the attos (10^-18). Because we wanted 1 AVAX == 1 ETH in the C-chain, that meant that AVAX needed to be more divisible on the C-chain.

File: vms/saevm/cchain/tx/compatibility_test.go (hunk 12 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/compatibility_test.go:65]: `AVAXAssetID` is the mainnet AVAX assetID (because we use mainnet txs as some of the test vectors)

File: vms/saevm/cchain/tx/tx.go (hunk 13 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/tx.go:92]: We will almost certainly never introduce another custom tx type.

File: vms/saevm/cchain/tx/export.go (hunk 19 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/export.go:12]: > Is there a linter for these doc comments.

Not that I know of

> When this package gets deleted this it will have to be fixed.

There will be a ton of stuff that breaks (in the tests and some constants) when we delete coreth. I think that's fine.

File: vms/saevm/cchain/tx/tx.go (hunk 18 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/tx.go:122]: This code needs to live in production alongside the `atomic` package for at least a little while. I'd prefer to have the references to where things came from rather than just magic numbers in the code...

Once coreth is removed, the single source of truth will be this package, so we wouldn't need the link anymore

File: vms/saevm/cchain/tx/tx.go (hunk 13 above)
Comment (JonathanOppenheimer) [reply vms/saevm/cchain/tx/tx.go:92]: Why do we refer to this all so generically then?

File: vms/saevm/cchain/tx/export.go (hunk 16 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/export.go:53]: So, I had originally looked into this... It's kind of annoying because each loop is its own type. So we could define a generic function or two with some anonymous functions... Personally I found it more gross than this code... But I totally understand wanting to clean it up... It's very similar code.

I think this is what it would look like if we wanted to share the code:
```golang
func (e *Export) burned(assetID ids.ID) (uint64, error) {
	return diffByAsset(
		assetID,
		e.Ins,
		func(i Input) (ids.ID, uint64) {
			return i.AssetID, i.Amount
		},
		e.ExportedOutputs,
		func(o *avax.TransferableOutput) (ids.ID, uint64) {
			return o.Asset.ID, o.Out.Amount()
		},
	)
}

func (i *Import) burned(assetID ids.ID) (uint64, error) {
	return diffByAsset(
		assetID,
		i.ImportedInputs,
		func(i *avax.TransferableInput) (ids.ID, uint64) {
			return i.Asset.ID, i.In.Amount()
		},
		i.Outs,
		func(o Output) (ids.ID, uint64) {
			return o.AssetID, o.Amount
		},
	)
}

func diffByAsset[I, O any](
	assetID ids.ID,
	ins []I,
	in func(I) (ids.ID, uint64),
	outs []O,
	out func(O) (ids.ID, uint64),
) (uint64, error) {
	consumed, err := sumByAsset(assetID, ins, in)
	if err != nil {
		return 0, err
	}
	produced, err := sumByAsset(assetID, outs, out)
	if err != nil {
		return 0, err
	}
	return math.Sub(consumed, produced)
}

func sumByAsset[T any](
	assetID ids.ID,
	items []T,
	get func(T) (ids.ID, uint64),
) (uint64, error) {
	var (
		total uint64
		err   error
	)
	for _, item := range items {
		if id, amount := get(item); id == assetID {
			total, err = math.Add(total, amount)
			if err != nil {
				return 0, err
			}
		}
	}
	return total, nil
}
```

While the current code looks like:
```golang
func (e *Export) burned(assetID ids.ID) (uint64, error) {
	var (
		burned uint64
		err    error
	)
	for _, in := range e.Ins {
		if in.AssetID == assetID {
			burned, err = math.Add(burned, in.Amount)
			if err != nil {
				return 0, err
			}
		}
	}
	for _, out := range e.ExportedOutputs {
		if out.Asset.ID == assetID {
			burned, err = math.Sub(burned, out.Out.Amount())
			if err != nil {
				return 0, err
			}
		}
	}
	return burned, nil
}

func (i *Import) burned(assetID ids.ID) (uint64, error) {
	var (
		burned uint64
		err    error
	)
	for _, in := range i.ImportedInputs {
		if in.Asset.ID == assetID {
			burned, err = math.Add(burned, in.In.Amount())
			if err != nil {
				return 0, err
			}
		}
	}
	for _, out := range i.Outs {
		if out.AssetID == assetID {
			burned, err = math.Sub(burned, out.Amount)
			if err != nil {
				return 0, err
			}
		}
	}
	return burned, nil
}
```

The implementation with the duplicate structure is less code and with less indirection... imo I think the duplication is better.

File: vms/saevm/cchain/tx/tx.go (hunk 15 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/tx.go:178]: sure

File: vms/saevm/cchain/tx/tx.go (hunk 17 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/tx.go:154]: Perhaps we will export `_x2cRate` as `X2CRate` at some point. I think we should avoid exporting the var, just so that we can't run into any global mutability issues... But if there is going to be a source of truth, I think it should be based on this value in this package. (or currently the C-chain equivalent).

But it feels out of scope of this PR.

Comment (JonathanOppenheimer) [reply vms/saevm/cchain/tx/tx.go:154]: Fair - there's a number of duplicated constants that perhaps can be cleaned after deletion.

File: vms/saevm/cchain/tx/export.go (hunk 16 above)
Comment (JonathanOppenheimer) [reply vms/saevm/cchain/tx/export.go:53]: Yeah I agree. The helper functions are worse I think.

File: vms/saevm/cchain/tx/tx.go (hunk 13 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/tx.go:92]: I don't really know what you mean. What is generic?

Comment (JonathanOppenheimer) [reply vms/saevm/cchain/tx/tx.go:92]: We don't ever call them atomic transactions - you could read all of this and be sort of blind to the fact that this is for atomic txs if you didn't already know - does that make sense or nah.

Comment (StephenButtolph) [reply vms/saevm/cchain/tx/tx.go:92]: I think we have generally moved away from calling them "Atomic" txs. In the X-chain and the P-chain, they are just ImportTx and ExportTx.

Coreth is really the only place we call them atomic txs... I feel like the name is pretty confusing as well there isn't really anything "atomic" about them.

File: vms/saevm/cchain/tx/tx_test.go (hunk 6 above)
Comment (JonathanOppenheimer) [reply vms/saevm/cchain/tx/tx_test.go:36]: this would go along well with a source of truth for these constants -- out of scope of the PR

File: vms/saevm/cchain/tx/tx_test.go
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/tx_test.go]: This stood out to me when looking at the file as being forgotten/not updated? might be out of scope.

File: vms/saevm/cchain/tx/tx_test.go [hunk 21]
```diff
@@ -133,8 +139,14 @@ var (
 					]
 				}]
 			}`,
-			ID:    ids.FromStringOrPanic("h34BPNmYApCbW8buVWAtzu1KtjTFmyMhiRQQnAqPqwCqQsB7f"),
 			Bytes: common.FromHex("0x000000000000000000010427d4b22a2a78bcddd456742caf91b56badbff985ee19aef14573e7343fd652ed5f38341e436e5d46e2bb00b45d62ae97d1b050c64bc634ae10626739e35c4b00000001c52b712aa7dce27a650bf509f799673e245edd4fa9e4e1700eb6105202fe579a0000000121e67317cbc4be2aeb00677ad6462778a8f52274b9d605df2591b23027a87dff000000050000000002faf080000000010000000000000001b8b5a87d1c05676f1f966da49151fa54dbe68c330000000002faf08021e67317cbc4be2aeb00677ad6462778a8f52274b9d605df2591b23027a87dff0000000100000009000000013e6614876ee01d3b8b27480c00bdcb0ae84ee3e8346d2d5f08320f7dd3e76c4540be021fe85e91817654c9310b54e8f2e88d81db52b8693842b90f3dbd23bd5c01"),
+			Op: hook.Op{
+				ID:  ids.FromStringOrPanic("h34BPNmYApCbW8buVWAtzu1KtjTFmyMhiRQQnAqPqwCqQsB7f"),
```
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/tx_test.go:144]: What's the value in hard-coding these per test? Does this catch regressions? 

Could we instead hash `test.Bytes` on the fly,to get the ID with some kind of helper for all of these instances -- I feel like that would be simpler than the current setup.

**Review by JonathanOppenheimer — APPROVED** [review]

File: vms/saevm/cchain/tx/tx.go (hunk 13 above)
Comment (JonathanOppenheimer) [reply vms/saevm/cchain/tx/tx.go:92]: Ah alright - I was unaware of that!

File: vms/saevm/cchain/tx/tx_test.go (hunk 21 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/tx_test.go:144]: I feel like hashing the bytes is just duplicating the code in the test
